core/lib/Slots.re

```diff
@@ -0,0 +1,83 @@
+type slotInternal('slot, 'nextSlots) = {
+  flush: unit => ('slot, bool),
+  this: 'slot,
+  next: 'nextSlots,
+};
+
+type t('slot, 'nextSlots) = ref(option(slotInternal('slot, 'nextSlots)));
+
+type empty = t(unit, unit);
+
+let create = () => ref(None);
+
+let flushPendingUpdatesInternal = (default, slots) =>
+  switch (slots) {
+  | Some({flush}) => flush()
+  | None => (default, false)
+  };
+
+let flushPendingUpdates = slots =>
+  switch (slots^) {
+  | Some({flush}) => snd(flush())
+  | None => false
+  };
+
+let shouldUpdate = (slots, newValue) =>
+  switch (slots) {
+  | Some({this}) => this !== newValue
+  | None => true
+  };
+
+let stabilized = (this, nextSlots) => {
+  this,
+  next: nextSlots,
+  flush: () => (this, false),
+};
+
+let makeSetter =
+    (~flushed, ~default, ~latestSlotsRef, ~nextSlots, makeNewValue) => {
+  let prevSlots = latestSlotsRef^;
+  latestSlotsRef :=
+    Some({
+      ...flushed,
+      flush: () => {
+        let prevValue = fst(flushPendingUpdatesInternal(default, prevSlots));
+        let newValue = makeNewValue(prevValue);
+        let shouldUpdate = shouldUpdate(prevSlots, newValue);
+        latestSlotsRef := Some(stabilized(newValue, flushed.next));
```

Comment:
@wokalski would it make sense to add the boolean as a third component of `slotInternal`, then avoid updating the `flush` function but instead update the value and boolean only?
So that the flush function is only created once, and there's no "do nothing" flush function.

I was also wondering whether the logic to traverse the slots recursively could be moved out of the flush function. So the flush function takes care of a single slot and there's separate logic to flush all the slots in turn. Maybe such generality is not needed, but it would open up the possibility to have more freedom in flushing strategies.

Reply:
I have tried to come up with a solution for that using diff lists. The idea is to deconstruct the input difflist/hooks and at the same time construct the output hooks which we will return from render. The input might be `None` and in that case we would construct the output difflist from the default values. That said, I cannot satisfy the type system here.

There are three types involved:
```
type index('full, 'nth) =
  | Zero: index('nth => 'more, 'nth)
  | Succ(index('full, 'nth)): index('more => 'full, 'nth);

type t('ty) =
  | Nil: t(empty)
  | Cons('a, t('ty)): t('a => 'ty);

let rec nth: type l v. (t(l), index(l, v)) => v
let dropFirst: type elem tail. t(elem => tail) => (elem, t(tail));

type t_appendable('contents, 'unificationVariable) =
  | Nil: t('unificationVariable, 'unificationVariable)
  | Cons('a, t('contents, 'uv)): t('a => 'contents, 'uv);

let rec addElement:
  type l elem uv uv2. (t_appendable(l, elem), t_appendable(elem, uv2)) => t_appendable(l, uv2)
```

Given those types I've been trying to come up with a sound implementation for `use*` function with the default argument which would both get `nth` element from the "input" list and append an element to the "output" list but I'm not sure if it's possible to express. One nice thing is that - together with making functorizing such lists, we could get a simple implementation of `iter` and less mutation.

Reply:
I'd like to note, that it's still possible to implement `iter`/`fold` on top of `Slots` (or at least I think so.)

Reply:
@cristianoc I have implemented folding/iteration on top of the current slots approach. As a bonus it should use less allocations because the `flush`/`enqueueUpdates`/`fold` functions are only created once. I'm still interested in using diff lists if it's possible but this approach is good enough.

I also functorized Slots so that we can make something more fun like creating such slots:

```
module Hooks = Slots.Make({
  type t('a) =
    | State('a)
    | Effect('a)
});
```

Thanks to that we will be able to flush state separately from the effects.